src/core/server/core_usage_data/core_usage_data_service.ts

```diff
@@ -493,7 +496,22 @@ export class CoreUsageDataService implements CoreService<CoreUsageDataSetup, Cor
 
     this.coreUsageStatsClient = getClient();
 
-    return { registerType, getClient } as CoreUsageDataSetup;
+    const contract: InternalCoreUsageDataSetup = {
+      registerType,
+      getClient,
+      registerUsageCounter: (usageCounter) => {
+        this.incrementUsageCounter = (params) => usageCounter.incrementCounter(params);
+      },
+      incrementUsageCounter: (params) => {
+        try {
+          this.incrementUsageCounter(params);
+        } catch (e) {
+          // Self-defense mechanism since the handler is externally registered
+        }
+      },
```

Comment:
> I would expect `incrementUsageCounter` to be only exposed on the `start` contract?

The `usageCounters` API exposes them in the `setup` contract. AFAIRecall, the reasoning is that we may want to report startup events (the plugin `usageCollection` holds them in memory until the SO is available and it can start dumping them to the corresponding SOs though). IMO, this is a good example: if `GET /api/status` is used as a health check, it will likely be called during the setup phase.

I'll cc @Bamieh (the original dev) for more insights.

> Also NIT: we may want to log an error if the call to the delegate fails.

Debug level? Warning seems too intrusive for a telemetry use case (that doesn't add direct value to the user). What do you think?

Reply:
> IMO, this is a good example: if GET /api/status is used as a health check, it will likely be called during the setup phase.

The usageCollection will likely not have called the register hook if the endpoint is hit during setup anyway?

But I do remember there were some egg-and-chiken condition for the saved objects endpoints counters.

> Debug level? Warning seems too intrusive for a telemetry use case

debug level is fine

Reply:
Correct the `incrementUsageCounter` function is exposed at setup. The data is stored in memory during `setup` then flushed during the `start` cycle into savedobjects once ES is ready.